Add Wallbox SMA EV Charger 22

powelllens opened the pull request:
Add Basic support for new Wallbox SMA EV Charger 22

    Requirements:
    Wallbox SMA EV Charger 22 - without SMA Home Manager (Settings would still be possible)

    Features:
    -Status
    -Enabled
    -Enable
    -MaxCurrent
    -MaxCurrentMillis
    -CurrentPower
    -ChargedEnergy
    -Currents

    Limitations:
    -need none Admin account to read data from Wallbox
    -Full Control only possible if switch is in "fast charge mode"

    ToDo:
    -Limited Error functionalities / bad feedback if something goes wrong
    -testing

    All Features have been tested.
    Functionality is given, would consider this a Beta.

powelllens committed:
Add files via upload

Add Basic support for new Wallbox SMA EV Charger 22

    Requirements:
    Wallbox SMA EV Charger 22 - without SMA Home Manager (Settings would still be possible)

    Features:
    -Status
    -Enabled
    -Enable
    -MaxCurrent
    -CurrentPower
    -ChargedEnergy
    -Currents

    Limitations:
    -need none Admin account to read data from Wallbox
    -Full Control only possible if switch is in "fast charge mode"

    ToDo:
    -Token is currently acquired during every read / write (bad practice - proper procedure in golang unknown)
    -MaxCurrent were programmed with hard lower and upper limits in source to prevent illegal use above 11kwh - could be changed
    -every read / write operation will trigger a full data dump from the wallbox
    -Limited Error functionalities / bad feedback if something goes wrong
    -no catch for timeouts if wall box is unavailable

    All Features have been tested.
    Functionality is given, would consider this a Beta.

powelllens commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,374 @@
+func (wb *Smaevcharger) MaxCurrent(current int64) error {
+
+	//Convert current to watts
+	ChargeRate := current * 3 * 230
```
Does not make sense for 1P/2P Loading, any suggestions?

int5749 replied:
at least 1P loading? as this is also used with SMA HM 2.0

powelllens replied:
I was unable to find any parameters that would enable a remote control of the used Phases. This seems to be an automated feature or this might be restricted to the SMA speedwire protocol. Indication of the used Phases could be done via the measured Current on each Phase as a  workaround.

andig replied:
Evcc has current+phases1p3p api. Both should- combined use- set the power and Sma should adjust the phases. Just an idea…

powelllens replied:
Found another option to Limit the MaxCurrent with the "EV Charger Supply" Current. I'll need to test different setups and charger behaviors. I'll probably test that with the [charger_ramp](https://github.com/evcc-io/evcc/blob/master/cmd/charger_ramp.go) as soon as i find some time for that. I'll check the resulting PWM Signal from the Charger and will post the findings.

powelllens replied:
add support for 10mA steps MaxCurrentMillis
changed parameters / behavior of current limiter
Tested Setup with EVCC ramp function and Renault:
![image](https://user-images.githubusercontent.com/6517959/165648125-8ce31245-b68b-495d-9e80-167912e43e18.png)

powelllens commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,374 @@
+		ChargeRate = 11000
+	} else if ChargeRate < 4100 {
+		ChargeRate = 4100
+	}
```
Hard coded Limits OK?

powelllens committed:
Merge pull request #1 from evcc-io/master

update to latest version

powelllens committed:
New Version of MaxCurrent for charger
- add support for 10mA steps MaxCurrentMillis
- changed parameters / behavior of current limiter

andig commented on charger/smaevcharger.go:
```diff
@@ -0,0 +1,405 @@
+
+//Constants
+const (
+	smaevchNConNCarNChar = float32(200111) // No Car connectec and no charging
```
Please move these….

powelllens committed:
-add auth  file /smaevcharger/identity.go
-changed Token Auth functionality Token will be used until expired, refresh of token via access token is currently not working. As a Workaround we acquire a new token with the username and password.

-moved const to /smaevcharger/types.go

-cleaned up several functions and used already existing helpers

-add multi parameter upload functionalities

-some minor changes and bugfixes

powelllens committed:
Merge branch 'evcc-io:master' into master

powelllens committed:
- changes Software to support Charger Software version 1.2.23R
- due to changed behavior earlier Software will not be supported
- add software version control during startup
- Current control workaround no longer needed, current is now be controlled with the AC Line Current Limiter

powelllens committed:
merged master

premultiply commented:
Rebase ?

andig commented:
@powelllens es gibt keinen Grund immer wider master rein zu mergen. Dein PR ist völlig autark- damit erzeugst Du schlimmstenfalls nur Konflikte.

powelllens commented:
I'll reopen a new PR as soon as I have my messed up branch sorted.
